cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java

```diff
@@ -271,12 +301,28 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
     nsStore.update(builder.build());
   }
 
-  private boolean checkProgramsRunning(final Id.Namespace namespaceId) {
+  private boolean checkProgramsRunning(final NamespaceId namespaceId) {
     return runtimeService.checkAnyRunning(new Predicate<Id.Program>() {
       @Override
       public boolean apply(Id.Program program) {
-        return program.getNamespaceId().equals(namespaceId.getId());
+        return program.getNamespaceId().equals(namespaceId.getNamespace());
       }
     }, ProgramType.values());
   }
+
+  private Principal getPrincipal() {
+    return new Principal(SecurityRequestContext.getUserId(), Principal.PrincipalType.USER);
+  }
+
+  private InstanceId createInstanceId(CConfiguration cConf) {
+    String instanceName = cConf.get(Constants.Security.Authorization.INSTANCE_NAME);
+    if (instanceName == null) {
+      instanceName = cConf.get(Constants.Router.ADDRESS);
```

Comment:
I'm not so sure about this approach. It needs some more thinking. I'll investigate and update this PR when I'm ready. Have addressed all other comments so far.